functioncall/functioncall.go

```diff
@@ -38,12 +39,21 @@ func (funcs Funcs) MarshalJSON() ([]byte, error) {
 
 func (params Params) MarshalJSON() ([]byte, error) {
 	required := []string{}
-	props := map[string]Param{}
+	props := map[string]any{}
 	for _, p := range params {
 		if p.Required {
 			required = append(required, p.Name)
 		}
-		props[p.Name] = p
+		if p.Type == "array" && p.Items != nil {
+			schema := map[string]any{
+				"type":     "array",
+				"items":    p.Items,
+				"required": required,
```

Comment:
I'm not sure why this `required` needs to be here. This `required` is supposed to contain all names of `required` params within this `Params`, not a `Param`. Could you give me more guidance here?

Reply:
Relatedly, do we need `if p.Items != nil` condition above?
If we don't need any special operation for `array` such as this `required`, I think `omitempty` for `Items` is enough.

Reply:
> I'm not sure why this `required` needs to be here. This `required` is supposed to contain all names of `required` params within this `Params`, not a `Param`. Could you give me more guidance here?我不确定为什么这个 `required` 需要在这里。这个 `required` 应该包含 `Params` 中 `required` 参数的所有名称，而不是 `Param` 。您能给我更多指导吗？

https://community.openai.com/t/how-to-calculate-the-tokens-when-using-function-call/266573/27
https://github.com/hmarr/openai-chat-tokens

```json

  {
    messages: [{ role: "user", content: "hello" }],
    functions: [
      {
        name: "get_recipe",
        parameters: {
          type: "object",
          required: ["ingredients", "instructions", "time_to_cook"],
          properties: {
            ingredients: {
              type: "array",
              items: {
                type: "object",
                required: ["name", "unit", "amount"],
                properties: {
                  name: {
                    type: "string",
                  },
                  unit: {
                    enum: ["grams", "ml", "cups", "pieces", "teaspoons"],
                    type: "string",
                  },
                  amount: {
                    type: "number",
                  },
                },
              },
            },
            instructions: {
              type: "array",
              items: {
                type: "string",
              },
              description: "Steps to prepare the recipe (no numbering)",
            },
            time_to_cook: {
              type: "number",
              description: "Total time to prepare the recipe in minutes",
            },
          },
        },
      },
    ],
    tokens: 106,
  },

Reply:
> Relatedly, do we need `if p.Items != nil` condition above? If we don't need any special operation for `array` such as this `required`, I think `omitempty` for `Items` is enough.

This is just my coding habit. Using only 'if p.Type == "array"' is also acceptable